### PR TITLE
Added crosspointer transmute error and tests 

### PR DESCRIPTION
**Pull request by cramertj**
Added lint for transmute between `T` and `*T` as specified in #749.

### CODE REVIEW

**Comment by mcarton** [comment]
You’ll need to run `python util/update_lints.py` for the tests to pass.

File: src/transmute.rs [hunk 1]
```diff
@@ -15,6 +17,19 @@ declare_lint! {
     "transmutes that have the same to and from types"
 }
 
+/// **What it does:*** This lint checks for transmutes between a type T and *T.
```
Comment (mcarton) [inline src/transmute.rs]: Can you put `T` and `*T` between ticks here? This will go to the markdown wiki.

File: src/transmute.rs [hunk 2]
```diff
@@ -43,3 +58,46 @@ impl LateLintPass for UselessTransmute {
...
+    }
+}
+
+impl LateLintPass for CrosspointerTransmute {
+    fn check_expr(&mut self, cx: &LateContext, e: &Expr) {
+        if let ExprCall(ref path_expr, ref args) = e.node {
+            if let ExprPath(None, _) = path_expr.node {
+                let def_id = cx.tcx.def_map.borrow()[&path_expr.id].def_id();
+
+                if utils::match_def_path(cx, def_id, &["core", "intrinsics", "transmute"]) {
```
Comment (mcarton) [inline src/transmute.rs]: You should put that path with the [big ugly list of other path](https://github.com/Manishearth/rust-clippy/blob/7e65493599d631f5be0fbc882974153bf85a5ad6/src/utils/mod.rs#L23-L55).

File: src/transmute.rs [hunk 3]
```diff
@@ -43,3 +58,46 @@ impl LateLintPass for UselessTransmute {
...
+                    if is_ptr_to(to_ty, from_ty) {
+                        cx.span_lint(CROSSPOINTER_TRANSMUTE,
+                                     e.span,
+                                     &format!("transmute from a type (`{}`) to a pointer to that type (`{}`)", from_ty, to_ty));
+                    }
+
+                    if is_ptr_to(from_ty, to_ty) {
+                        cx.span_lint(CROSSPOINTER_TRANSMUTE,
+                                     e.span,
+                                     &format!("transmute from a type (`{}`) to the type that it points to (`{}`)", from_ty, to_ty));
```
Comment (mcarton) [inline src/transmute.rs]: One of those should probably have the `from_ty, to_ty` parameter reversed.

**Comment by mcarton** [comment]
Looks good modulo some nits :smile:

File: src/transmute.rs [hunk 4]
```diff
@@ -43,3 +58,46 @@ impl LateLintPass for UselessTransmute {
...
+                    let from_ty = cx.tcx.expr_ty(&args[0]);
+                    let to_ty = cx.tcx.expr_ty(e);
+
+                    if is_ptr_to(to_ty, from_ty) {
+                        cx.span_lint(CROSSPOINTER_TRANSMUTE,
+                                     e.span,
+                                     &format!("transmute from a type (`{}`) to a pointer to that type (`{}`)", from_ty, to_ty));
+                    }
+
+                    if is_ptr_to(from_ty, to_ty) {
```
Comment (mcarton) [inline src/transmute.rs]: That one can be an `else if`.

**Comment by llogiq** [comment]
Thanks!

**Comment by cramertj** [comment]
@mcarton Thanks for your help with my first clippy PR!

**Pull request merged**